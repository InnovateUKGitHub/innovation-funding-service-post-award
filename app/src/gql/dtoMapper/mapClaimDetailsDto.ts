import type { ClaimDetailsDto } from "@framework/dtos/claimDetailsDto";
import { mapImpactManagementParticipationToEnum } from "@framework/mappers/impactManagementParticipation";
import { Clock, salesforceDateFormat } from "@framework/util/clock";

const clock = new Clock();

type ClaimDetailsNode = Readonly<
  Partial<{
    Acc_ClaimStatus__c: GQL.Value<string>;
    Acc_CostCategory__c: GQL.Value<string>;
    Acc_PeriodCostCategoryTotal__c: GQL.Value<number>;
    Acc_ProjectPeriodEndDate__c: GQL.Value<string>;
    Acc_ProjectPeriodNumber__c: GQL.Value<number>;
    Acc_ProjectPeriodStartDate__c?: GQL.Value<string>;
    Acc_Grant_Paid_To_Date__c: GQL.Value<number>;
    Impact_Management_Participation__c: GQL.Value<string>;
    RecordType: {
      Name: GQL.Value<string>;
    } | null;
  }>
> | null;

type ClaimDetailsDtoMapping = Pick<
  ClaimDetailsDto,
  | "periodId"
  | "costCategoryId"
  | "value"
  | "periodEnd"
  | "periodStart"
  | "impactManagementParticipation"
  | "grantPaidToDate"
>;

const mapper: GQL.DtoMapper<ClaimDetailsDtoMapping, ClaimDetailsNode> = {
  costCategoryId: function (node) {
    return node?.Acc_CostCategory__c?.value ?? "unknown";
  },
  periodEnd: function (node) {
    return clock.parse(node?.Acc_ProjectPeriodEndDate__c?.value ?? null, salesforceDateFormat);
  },
  periodId: function (node) {
    return node?.Acc_ProjectPeriodNumber__c?.value ?? 0;
  },
  periodStart: function (node) {
    return clock.parse(node?.Acc_ProjectPeriodStartDate__c?.value ?? null, salesforceDateFormat);
  },
  value: function (node) {
    return node?.Acc_PeriodCostCategoryTotal__c?.value ?? 0;
  },
  impactManagementParticipation: function (node) {
    return mapImpactManagementParticipationToEnum(node?.Impact_Management_Participation__c?.value);
  },
  grantPaidToDate(node) {
    return node?.Acc_Grant_Paid_To_Date__c?.value ?? 0;
  },
};

/**
 * Maps a specified ClaimDetails Node from a GQL query to a slice of
 * the ClaimDetailsDto to ensure consistency and compatibility in the application
 */
export function mapToClaimDetailsDto<T extends ClaimDetailsNode, PickList extends keyof ClaimDetailsDtoMapping>(
  node: T,
  pickList: PickList[],
): Pick<ClaimDetailsDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](node);
    return dto;
  }, {} as Pick<ClaimDetailsDtoMapping, PickList>);
}

/**
 * Maps ClaimDetails Edge to array of ClaimDetails DTOs.
 * It is filtered for correct Record Type of "Claim Detail"
 */
export function mapToClaimDetailsDtoArray<
  T extends ReadonlyArray<{ node: ClaimDetailsNode } | null> | null,
  PickList extends keyof ClaimDetailsDtoMapping,
>(edges: T, pickList: PickList[]): Pick<ClaimDetailsDtoMapping, PickList>[] {
  return (
    edges
      ?.filter(
        x =>
          x?.node?.RecordType?.Name?.value === "Claims Detail" &&
          x?.node?.Acc_ClaimStatus__c?.value !== "New" &&
          x?.node?.Acc_CostCategory__c?.value !== null,
      )
      ?.map(node => {
        return mapToClaimDetailsDto(node?.node ?? null, pickList);
      }) ?? []
  );
}
