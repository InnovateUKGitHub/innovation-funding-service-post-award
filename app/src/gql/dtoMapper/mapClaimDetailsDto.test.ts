import { mapToClaimDetailsDtoArray } from "./mapClaimDetailsDto";

describe("mapToClaimDetailsArray", () => {
  const edges = [
    {
      node: {
        Acc_CostCategory__c: {
          value: "category-id-0",
        },
        Acc_ProjectPeriodStartDate__c: {
          value: "2020-01-21",
        },
        Acc_ProjectPeriodEndDate__c: {
          value: "2020-03-21",
        },
        Acc_ProjectPeriodNumber__c: {
          value: 2,
        },
        Acc_PeriodCostCategoryTotal__c: {
          value: 10000,
        },
        Acc_ClaimStatus__c: {
          value: "Current",
        },
        RecordType: {
          DeveloperName: {
            value: "Claims_Detail",
          },
        },
      },
    },
    {
      node: {
        Acc_CostCategory__c: {
          value: "category-id-1",
        },
        Acc_ProjectPeriodStartDate__c: {
          value: "2020-01-21",
        },
        Acc_ProjectPeriodEndDate__c: {
          value: "2020-03-21",
        },
        Acc_ProjectPeriodNumber__c: {
          value: 3,
        },
        Acc_PeriodCostCategoryTotal__c: {
          value: 10000,
        },
        Acc_ClaimStatus__c: {
          value: "Current",
        },
        RecordType: {
          DeveloperName: {
            value: "Claims_Detail",
          },
        },
      },
    },
    {
      node: {
        Acc_CostCategory__c: {
          value: "category-defaults",
        },
        Acc_ProjectPeriodStartDate__c: null,
        Acc_ProjectPeriodEndDate__c: null,
        Acc_ProjectPeriodNumber__c: null,
        Acc_PeriodCostCategoryTotal__c: null,
        Acc_ClaimStatus__c: {
          value: "Current",
        },
        RecordType: {
          DeveloperName: {
            value: "Claims_Detail",
          },
        },
      },
    },
    {
      node: {
        Acc_CostCategory__c: {
          value: null,
        },
        Acc_ProjectPeriodStartDate__c: null,
        Acc_ProjectPeriodEndDate__c: null,
        Acc_ProjectPeriodNumber__c: null,
        Acc_PeriodCostCategoryTotal__c: null,
        Acc_ClaimStatus__c: {
          value: "Current",
        },
        RecordType: {
          DeveloperName: {
            value: "Claims_Detail",
          },
        },
      },
    },
    {
      node: {
        Acc_CostCategory__c: {
          value: "category-id-1",
        },
        Acc_ProjectPeriodStartDate__c: {
          value: "2020-01-21",
        },
        Acc_ProjectPeriodEndDate__c: {
          value: "2020-03-21",
        },
        Acc_ProjectPeriodNumber__c: {
          value: 3,
        },
        Acc_PeriodCostCategoryTotal__c: {
          value: 10000,
        },
        Acc_ClaimStatus__c: {
          value: "New",
        },
        RecordType: {
          DeveloperName: {
            value: "Claims_Detail",
          },
        },
      },
    },
    {
      node: {
        Acc_CostCategory__c: {
          value: "category-id-1",
        },
        Acc_ProjectPeriodStartDate__c: {
          value: "2020-01-21",
        },
        Acc_ProjectPeriodEndDate__c: {
          value: "2020-03-21",
        },
        Acc_ProjectPeriodNumber__c: {
          value: 3,
        },
        Acc_PeriodCostCategoryTotal__c: {
          value: 10000,
        },
        Acc_ClaimStatus__c: {
          value: "Current",
        },
        RecordType: {
          DeveloperName: {
            value: "Random",
          },
        },
      },
    },
  ];

  it("should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records", () => {
    expect(
      mapToClaimDetailsDtoArray(edges, ["costCategoryId", "periodEnd", "periodId", "periodStart", "value"], {}),
    ).toMatchSnapshot();
  });

  it("should map the gql data only including fields on the picklist", () => {
    expect(mapToClaimDetailsDtoArray(edges, ["costCategoryId", "value"], {})).toMatchSnapshot();
  });
});
