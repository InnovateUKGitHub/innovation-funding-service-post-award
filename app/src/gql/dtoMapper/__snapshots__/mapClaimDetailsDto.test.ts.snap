// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToClaimDetailsArray should map the gql data only including fields on the picklist 1`] = `
[
  {
    "costCategoryId": "category-id-0",
    "value": 10000,
  },
  {
    "costCategoryId": "category-id-1",
    "value": 10000,
  },
  {
    "costCategoryId": "category-defaults",
    "value": 0,
  },
]
`;

exports[`mapToClaimDetailsArray should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records 1`] = `
[
  {
    "costCategoryId": "category-id-0",
    "periodEnd": 2020-03-21T00:00:00.000Z,
    "periodId": 2,
    "periodStart": 2020-01-21T00:00:00.000Z,
    "value": 10000,
  },
  {
    "costCategoryId": "category-id-1",
    "periodEnd": 2020-03-21T00:00:00.000Z,
    "periodId": 3,
    "periodStart": 2020-01-21T00:00:00.000Z,
    "value": 10000,
  },
  {
    "costCategoryId": "category-defaults",
    "periodEnd": null,
    "periodId": 0,
    "periodStart": null,
    "value": 0,
  },
]
`;
